### ISSUE TITLE
Positional is treated as a sub-command

### ISSUE THREAD

**Issue by gon1332** [henryiii/CLI11#1022]
CLI11: Version 2.4.1

I have this setup:

```c++
int main() {
    CLI::App cli{};
    cli.set_help_flag();
    cli.add_subcommand("foo", "Does foo")->parse_complete_callback([&](){std::cout << "=== Foo doing foo" << std::endl;});
    cli.add_subcommand("bar", "Does bar")->parse_complete_callback([&](){std::cout << "=== Bar doing bar" << std::endl;});
    auto *help_cmd = cli.add_subcommand("help", "Does help");
    std::string help_for_command;
    auto *help_opt = help_cmd->add_option("command", help_for_command, "Command to print help for")->expected(0, 1);
    help_cmd->parse_complete_callback([&]() {
            if (*help_opt) {
                std::cout << "Print help for " << help_for_command << std::endl;
            } else {
                CLI::Formatter fmt{};
                std::cout << fmt.make_help(&cli, "", CLI::AppFormatMode::Normal) << std::endl;
            }
    });

    std::cout << "=> " << std::flush;
    for (std::string line; std::getline(std::cin, line);) {
        try {
            cli.parse(line);
        } catch (const CLI::ParseError &e) {
            std::cout << e.what() << std::endl;
            std::cout << cli.help();
        }
        std::cout << "=> " << std::flush;
    }
}
```

Given the above setup, here are some behaviors that I expect:
```
Scenario 1
When help is called without arguments
Then the help string of all commands is printed

Scenario 2
When help is called with one argument
And   the argument is a registered (sub)command
Then the help string of this command is printed

Scenario 3
When help is called with one argument
And   the argument is not a registered (sub)command
Then the help string of all commands is printed
```

However, I get the below output for each scenario:

Scenario 1 🟢 
```shell
=>help
 help
Usage: [SUBCOMMAND]

Subcommands:
  foo                         Does foo
  bar                         Does bar
  help                        Does help

=>
```

Scenario 2 🔴 
Not only the positional is not recognised, it is executed as a sucommand.
```shell
=>help foo
 help foo
Usage: [SUBCOMMAND]

Subcommands:
  foo                         Does foo
  bar                         Does bar
  help                        Does help

=== Foo doing foo
=>
```

```shell
=>help help
 help help
Usage: [SUBCOMMAND]

Subcommands:
  foo                         Does foo
  bar                         Does bar
  help                        Does help

Usage: [SUBCOMMAND]

Subcommands:
  foo                         Does foo
  bar                         Does bar
  help                        Does help

=>
```

Scenario 3 🔴 
```shell
=>help other
 help other
Usage: [SUBCOMMAND]

Subcommands:
  foo                         Does foo
  bar                         Does bar
  help                        Does help

The following argument was not expected: other
Does help
Usage: help [command]

Positionals:
  command TEXT                Command to print help for
```

I've tried using `fallthrough(false)` for `help` command, but with no luck.
How could I implement my use case?

**Comment by phlptp** [comment]
I think what you want here is to use `prefix_command()` on the help subcommand.  That will force everything that comes after it to be a positional argument.  

Otherwise subcommands would take precedence over positionals

**Comment by gon1332** [comment]
Thanks for your timely response.

I've tried it as such, but it didn't help:

```c++
int main() {
    CLI::App cli{};
    cli.set_help_flag();
    cli.add_subcommand("foo", "Does foo")->parse_complete_callback([&](){std::cout << "=== Foo doing foo" << std::endl;});
    cli.add_subcommand("bar", "Does bar")->parse_complete_callback([&](){std::cout << "=== Bar doing bar" << std::endl;});
    auto *help_cmd = cli.add_subcommand("help", "Does help");
    std::string help_for_command;
    auto *help_opt = help_cmd->add_option("command", help_for_command, "Command to print help for")->expected(0, 1);
    help_cmd
        ->prefix_command()
        ->parse_complete_callback([&]() {
            if (*help_opt) {
                std::cout << "Print help for " << help_for_command << std::endl;
            } else {
                CLI::Formatter fmt{};
                std::cout << fmt.make_help(&cli, "", CLI::AppFormatMode::Normal) << std::endl;
            }
    });
    ...
}
```

Is there anything that I'm missing on the above snippet?

**Comment by phlptp** [comment]
I played with it a bit.   There does seem to be a gap in configuration capability for this scenario.
I believe this code does what you want
```C++
int main() {
    CLI::App cli{};
    cli.set_help_flag();
    cli.add_subcommand("foo", "Does foo")->parse_complete_callback([&](){std::cout << "=== Foo doing foo" << std::endl;});
    cli.add_subcommand("bar", "Does bar")->parse_complete_callback([&](){std::cout << "=== Bar doing bar" << std::endl;});
    auto *help_cmd = cli.add_subcommand("help", "Does help");
    std::string help_for_command;
    auto *help_opt = help_cmd->add_option("command", help_for_command, "Command to print help for");
    help_cmd
        ->preparse_callback([&](size_t values){help_opt->required(values>0);})
        ->parse_complete_callback([&]() {
        if (*help_opt) {
            std::cout << "Print help for " << help_for_command << std::endl;
        } else {
            CLI::Formatter fmt{};
            std::cout << fmt.make_help(&cli, "", CLI::AppFormatMode::Normal) << std::endl;
        }
            });
    std::cout << "=> " << std::flush;
    for (std::string line; std::getline(std::cin, line);) {
        try {
            cli.parse(line);
        } catch (const CLI::ParseError &e) {
            std::cout << e.what() << std::endl;
            std::cout << cli.help();
        }
        std::cout << "=> " << std::flush;
    }
    return 0;
}
```
but it isn't obvious, so there might be a missing option on subcommands that needs to get added for situations like this.  The use for prefix_command is slightly different, so doesn't work in this case.  What the above code does is take advantage of the preparse callback and the fact that required positionals take precedence over subcommands.

**Comment by gon1332** [comment]
Thank you very much. That seemed to do the trick.

An option that does this exact thing would be quite handy.

**Comment by phlptp** [comment]
This will be resolved in #1073

``` c++
int main() {
    CLI::App cli{};
    cli.set_help_flag();
    cli.add_subcommand("foo", "Does foo")->parse_complete_callback([&](){std::cout << "=== Foo doing foo" << std::endl;});
    cli.add_subcommand("bar", "Does bar")->parse_complete_callback([&](){std::cout << "=== Bar doing bar" << std::endl;});
    auto *help_cmd = cli.add_subcommand("help", "Does help")->subcommand_fallthrough(false);
    std::string help_for_command;
    auto *help_opt = help_cmd->add_option("command", help_for_command, "Command to print help for");
    help_cmd
        ->parse_complete_callback([&]() {
        if (*help_opt) {
            std::cout << "Print help for " << help_for_command << std::endl;
        } else {
            CLI::Formatter fmt{};
            std::cout << fmt.make_help(&cli, "", CLI::AppFormatMode::Normal) << std::endl;
        }
            });
    std::cout << "=> " << std::flush;
    for (std::string line; std::getline(std::cin, line);) {
        try {
            cli.parse(line);
        } catch (const CLI::ParseError &e) {
            std::cout << e.what() << std::endl;
            std::cout << cli.help();
        }
        std::cout << "=> " << std::flush;
    }
    return 0;
}
```
